catkit2/services/thorlabs_cube_motor_kinesis/thorlabs_cube_motor_kinesis.py

```diff
@@ -0,0 +1,233 @@
+            self.log.debug("Device parameters don't match configuration parameters %r %r %r.",
+                           self.min_position_device < self.min_position_config,
+                           self.max_position_device > self.max_position_config,
+                           self.unit != unit_config)
+            raise ValueError("Device parameters don't match configuration parameters.")
+
+        self.command = self.make_data_stream('command', 'float64', [1], 20)
+        self.current_position = self.make_data_stream('current_position', 'float64', [1], 20)
+
+        # Submit motor starting position to current_position data stream
+        self.get_current_position()
+
+        self.make_command('home', self.home)
+
+    def main(self):
+        while not self.should_shut_down:
+            try:
+                # Get an update for the motor position.
+                frame = self.command.get_next_frame(10)
+                # Set the current position if a new command has arrived.
+                self.set_current_position(frame.data[0])
+            except RuntimeError:
+                # Timed out. This is used to periodically check the shutdown flag.
+                continue
+
+    def close(self):
+        self.lib.CC_ClearMessageQueue(self.serial_number)
+        self.lib.CC_StopPolling(self.serial_number)
+        # Close the device
+        self.lib.CC_Close(self.serial_number)
+
+    def set_current_position(self, position):
+        """
+        Set the motor's current position.
+
+        The unit is given in real-life units (mm if translation, deg if rotation).
+        """
+        if self.min_position_config <= position <= self.max_position_config:
+            new_pos_real = c_double(position)  # in real units
+            new_pos_dev = c_int()
+            self.lib.CC_GetDeviceUnitFromRealValue(self.serial_number,
+                                                   new_pos_real,
+                                                   byref(new_pos_dev),
+                                                   0)
+            self.lib.CC_MoveToPosition(self.serial_number, new_pos_dev)
+        else:
+            self.log.warning('Motor not moving since commanded position is outside of configured range.')
+            self.log.warning('Position limits: %f %s <= position <= %f %s.',
+                             self.min_position_config, self.unit, self.max_position_config, self.unit)
+
+        # Update the current position data stream.
+        self.get_current_position()
+
+    def get_current_position(self):
+        current_position = self.lib.CC_GetPosition(self.serial_number)
+        real_unit = c_double()
+        self.lib.CC_GetRealValueFromDeviceUnit(self.serial_number, current_position, byref(real_unit), 0)
+        self.log.debug("Current position: %f %s", real_unit.value, self.unit)
+        self.current_position.submit_data(np.array([real_unit.value], dtype='float64'))
+
+    def wait_for_completion(self):

```

Comment:
The motor itself behaves differently when homing and when moving position, which is why we didn't block it with this wait when it's just changing position. Let me remind myself of the details and then either justify our choice or follow your advice and put it in there as well.

Reply:
Ok this one is tricky. If we do not wait for completion in the `set_current_position()`, it intercepts the first command and executes the second command instead, and correctly. So I am fine with this behavior and don't want to add the wait_for_completion there.

However, the method `get_current_position()` gets called too early and puts an intermediate position on the current_position data stream, which the device reads while it is moving. Which is obviously not at all what we want.

We then tried adding the `wait_for_completion()` anyway, to fix the wrong values on the data stream, but in this case it just finishes running the first command. Data stream value is still wrong. *And* it puts command 2 in a cache apparently and runs it before the next command that is sent to the motor... also not the behavior we want, so I still vote against adding the `wait_for_completion()` in `set_current_position()`.

Since this was so weird, we checked the data stream values after calling `home()` a couple of times and this is always fine, so it's something about the command data stream monitoring and getting the real device position before the device stops moving.

I don't actually understand what `wait_for_completion()` does. @a-sevin, maybe you can help out with this conundrum? What does it do and what are the values for `message_id` and `message_type` it checks against?

The minimum requirement for this fix is that the correct (final) value is put on the current_position data stream after the movement concludes.

Reply:
Maybe I should just regularly update the current position, like the Newport XPS service?
https://github.com/spacetelescope/catkit2/blob/6c8900abfd9f96e232a6a9749eaeee2760a5dbbe/catkit2/services/newport_xps_q8/newport_xps_q8.py#L191

@raphaelpclt what do you thunk about this?

Reply:
I'm reading the tutorial labview is not so bad, I need to check what to use it in Python:
https://www.thorlabs.com/Software/Motion%20Control/Kinesis/Kinesis-labview.pdf

This `wait_for_completion()` could be very simple: wait the `get_current_position()` stable...

Reply:
For the `get_current_position`, yes I think the solution with the newport xps is what you want. Then I am not sure of what you want for the moving behavior. The fact that the moving function returns before completion sounds a bit weird to me, but maybe not dramatic.

Reply:
Considering we now get correct values on the current_position data stream within the time scales we need (1 s), I will open a follow-up issue to look into a more rigorous handling of the data stream update inside of `set_current_position()`.

Issue: https://halo.sciencesconf.org/?lang=en

Come :) They can also pay for non-permanents.